### PR TITLE
Add step closures

### PR DESCRIPTION
**Pull request by antoniojkim**
Ports over the step closure functionality from PyTorch/XLA to Lazy Tensor Core:

References:
https://github.com/pytorch/xla/blob/205ae574c0a24e092899ea8610c360f93f5d8142/torch_xla/core/xla_model.py#L852-L900
https://github.com/pytorch/xla/blob/205ae574c0a24e092899ea8610c360f93f5d8142/torch_xla/utils/closures.py#L7-L83

CC: @wconstab @JackCaoG @Krovatkin

### CODE REVIEW

File: torch/_lazy/__init__.py [hunk 1]
```diff
@@ -1,6 +1,10 @@
+import threading
+
 import torch._C._lazy
 from torch.utils._pytree import tree_flatten, tree_unflatten
 
+from .closure import add_step_closure, run_step_closures
```
Comment (JackCaoG) [inline torch/_lazy/__init__.py:6]: I guess you don't need `add_step_closure` and `threading` import here?

File: torch/_lazy/closure.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        setattr(devctx, closures_type, step_closures)
+    step_closures.append(lambda a=args: closure(*a))
+
+
+def run_step_closures():
+    devctx = get_device_context()
+    async_step_closures = getattr(devctx, "async_step_closures", None)
+    if async_step_closures is not None:
+        devctx.async_step_closures = []
+        async_closure_handler = getattr(devctx, "async_closure_handler", None)
```
Comment (JackCaoG) [inline torch/_lazy/closure.py:120]: I don't see a `async_closure_handler` in `devctx`

Comment (JackCaoG) [reply torch/_lazy/closure.py:120]: nvm, it is set below

**Review by JackCaoG — APPROVED** [review]
It is the same step closure code in pytorch/xla, it has been working well so lgtm

**Comment by JackCaoG** [comment]
well, I guess some of the changes here is outside of my approval filed according to https://github.com/pytorch/pytorch/blob/d648375f13b4a4efd4cd35247098679fce5d4bcd/.github/merge_rules.yaml#L303.

**Review by wconstab — APPROVED** [review]

**Pull request closed without merging**